packages/flutter_tools/lib/src/base/error_handling_io.dart

```diff
@@ -272,6 +274,61 @@ class ErrorHandlingFile
     );
   }
 
+  /// This copy method attempts to handle file system errors from both reading
+  /// and writing the copied file.
+  @override
+  File copySync(String newPath) {
+    final File resultFile = fileSystem.file(newPath);
+    // First check if the source file can be read. If not, bail through error
+    // handling.
+    _runSync<void>(
+      () => delegate.openSync(mode: FileMode.read).closeSync(),
+      platform: _platform,
+      failureMessage: 'Flutter failed to copy $path to $newPath due to source location error'
+    );
+    // Next check if the destination file can be written. If not, bail through
+    // error handling.
+    _runSync<void>(
+      ()  {
+        resultFile.createSync(recursive: true);
+        resultFile.openSync(mode: FileMode.writeOnly).closeSync();
+      },
+      platform: _platform,
+      failureMessage: 'Flutter faiuled to copy $path to $newPath due to destination location error'
+    );
+    // If both of the above checks passed, attempt to copy the file and catch
+    // any thrown errors.
+    try {
+      return wrapFile(delegate.copySync(newPath));
+    } on FileSystemException {
+      // Proceed below
+    }
+    // If the copy failed but both of the above checks passed, copy the bytes
+    // directly.
```

Comment:
If something goes wrong with this, we should probably delete the destination file.

Reply:
done